tests/contest/runtimetest/src/tests.rs

```diff
@@ -381,6 +384,114 @@ pub fn validate_scheduler_policy(spec: &Spec) {
     }
 }
 
+pub fn validate_devices(spec: &Spec) {
+    let linux = spec.linux().as_ref().unwrap();
+    if let Some(devices) = linux.devices() {
+        for (i, device) in devices.iter().enumerate() {
+            validate_device(
+                device,
+                &format!(
+                    "{} (linux.devices[{}])",
+                    device.path().as_path().to_str().unwrap(),
+                    i
+                ),
+            );
+        }
+    }
+}
+
+fn validate_device(device: &LinuxDevice, description: &str) {
+    let file_data = match fs::metadata(device.path()) {
+        Ok(data) => data,
+        Err(e) => {
+            if e.kind() == std::io::ErrorKind::NotFound {
+                return;
+            }
+            eprintln!(
+                "error due to fail to get metadata for device path {:?}, error: {}",
+                device.path(),
+                e
+            );
+            return;
+        }
+    };
+
+    let mut expected_type = device.typ();
+    if expected_type == LinuxDeviceType::U {
+        expected_type = LinuxDeviceType::C;
+    }
+
+    let file_type = file_data.file_type();
+    let actual_type = if file_type.is_char_device() {
+        LinuxDeviceType::C
+    } else if file_type.is_block_device() {
+        LinuxDeviceType::B
+    } else if file_type.is_fifo() {
+        LinuxDeviceType::P
+    } else {
+        LinuxDeviceType::U
+    };
+
+    if actual_type != expected_type {
+        eprintln!("error due to device type want {expected_type:?}, got {actual_type:?}");
+    }
+
+    if actual_type != LinuxDeviceType::P {
+        let dev = file_data.st_rdev();
+        let major = (dev >> 8) & 0xfff;
+        let minor = (dev & 0xff) | ((dev >> 12) & 0xfff00);
+        if major != device.major() as u64 {
+            eprintln!(
+                "error due to device major want {}, got {}",
+                device.major(),
+                major
+            );
+        }
+        if minor != device.minor() as u64 {
+            eprintln!(
+                "error due to device minor want {}, got {}",
+                device.minor(),
+                minor
+            );
+        }
+    }
+
+    let actual_permissions = file_data.permissions().mode() & 0o777;
+    let expected_permissions = device.file_mode();
+    if expected_permissions.is_none() {
+        return;
+    }
```

Comment:
We should probably get the actual permissions after checking if expected permissions are set